manim/utils/hashing.py

```diff
@@ -30,56 +32,132 @@ def default(self, obj):
+            return repr(obj)
         elif hasattr(obj, "__dict__"):
             temp = getattr(obj, "__dict__")
-            return self._encode_dict(temp)
+            # MappingProxy is not supported by the Json Encoder
+            if isinstance(temp, MappingProxyType):
+                return dict(temp)
+            return self._check_iterable(temp)
         elif isinstance(obj, np.uint8):
             return int(obj)
-        try:
-            return json.JSONEncoder.default(self, obj)
-        except TypeError:
-            # This is used when the user enters an unknown type in CONFIG. Rather than throwing an error, we transform
-            # it into a string "Unsupported type for hashing" so that it won't affect the hash.
-            return "Unsupported type for hashing"
-
-    def _encode_dict(self, obj):
-        """Clean dicts to be serialized : As dict keys must be of the type (str, int, float, bool), we have to change them when they are not of the right type.
-        To do that, if one is not of the good type we turn it into its hash using the same
-        method as all the objects here.
+
+        return f"Unsupported type for serializing -> {str(type(obj))}"
+
+    def _handle_already_processed(self, obj):
+        """Handle if an object has been already processed by checking the id of the object.
+
+        This prevents the mechanism to handle an object several times, and is used to prevent any circular reference.
 
         Parameters
         ----------
         obj : Any
-            The obj to be cleaned.
+            The obj to check.
 
         Returns
         -------
         Any
-            The object cleaned following the processus above.
+            "already_processed" string if it has been processed, otherwise obj.
         """
+        global ALREADY_PROCESSED_ID
+        if id(obj) in ALREADY_PROCESSED_ID:
+            return "already_processed"
+        if not isinstance(obj, (str, int, bool, float)):
+            ALREADY_PROCESSED_ID[id(obj)] = obj
+        return obj
 
-        def key_to_hash(key):
-            if not isinstance(key, (str, int, float, bool)) and key is not None:
-                # print('called')
-                return zlib.crc32(json.dumps(key, cls=CustomEncoder).encode())
-            return key
+    def _check_iterable(self, iterable):
+        """Check for circular reference at each iterable that will go through the JSONEncoder, as well as key of the wrong format.
 
-        if isinstance(obj, dict):
-            return {key_to_hash(k): self._encode_dict(v) for k, v in obj.items()}
-        return obj
+        If a key with a bad format is found (i.e not a int, string, or float), it gets replaced byt its hash using the same process implemented here.
+        If a circular reference is found within the iterable, it will be replaced by the string "already processed".
+
+        Parameters:
+        -----------
```

Comment:
Now sing with me

No colon
No colon
No colon
No colon
hell yea
```suggestion
        Parameters
        ----------
```

Reply:
We really need to find a tool that does this autoformatting for us.

Reply:
Like black but for docstrings